File: vigenere.c
/**
 * vigenere.c
 *
 * Scrambles a message using vigenere cipher
 *
 * Usage: ./vigenere k
 *
 * where k is a single keyword of alphbetical characters
 */

#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>

int main(int argc, string argv[])
{
    // Print error message and quit if user didn't enter the correct number fo command line arguments
    if(argc != 2)
    {
        printf("Usage: ./vigenere k\n");
        return 1;
    }
    
    string key = argv[1];
    
    // Loop over characters to make sure that they're alphabetical
    for(int i=0, n = strlen(key); i < n; i++)
    {
        if(!isalpha(key[i]))
        {
            printf("Usage: ./vigenere k\n");
            return 1;
        }
    }
 
    printf("plaintext:");
    string p = get_string();
    printf("ciphertext:");
    
    int shift;
    int keyLength = strlen(key);
    
    // Loop over chartacters in plaintext
    for (int i = 0, n = strlen(p), j = 0; i < n; i++)
    {
        // Get alphabetical index of the j'th character in the key
        if(isupper(key[j%keyLength]))
        {
            shift = key[j%keyLength] - 'A';
        }
        else
        {
            shift = key[j%keyLength] - 'a';    
        }
        
        // Check if i'th character in p is alphabetical
        if(isalpha(p[i]))
        {
           // Determine the case of the letter and handle accordingly
            if(isupper(p[i]))
            {
                // Convert the character index from ASCII to alphbatized. Shift it by the key and print the result 
                printf("%c", (p[i] - 'A' + shift) % 26 + 'A');
                j++;
                 
            }
            else
            {
                printf("%c", (p[i]- 'a' + shift) % 26 + 'a');
                j++;
                 
            }
        }
        // If character isn't alphabetical, go ahead and print it out as is
        else
        {
            printf("%c", p[i]);
        }
    }

    printf("\n");
}

File: vigenere.py
"""
vigenere.py
Usage: vigenere.py k
Where k is a key containaing only aplphabetical character

Luke Benz
"""
import cs50
import sys
    
# Ensure proper usage
if len(sys.argv) != 2:
    sys.exit("Usage: vigenere.py k")
    
# Store key and ensure all characters are alphabetical
key = sys.argv[1]
for i in range(len(key)):
    if str.isalpha(key[i]) == False:
        sys.exit("Usage: vigenere.py k")
        
print("plaintext: ", end = "")
message = cs50.get_string()
print("ciphertext: ", end = "")
    
shift = 0
counter = 0
keyLength = len(key)
    
# Loop over characters in plaintext
for i in range(len(message)):
    # Get alphabetical index of the counter character in the key
    if str.isupper(key[counter%keyLength]):
        shift = ord(key[counter%keyLength]) - ord('A')
    else:
        shift = ord(key[counter%keyLength]) - ord('a')
        
    # Check if i'th character in message is alphabetic
    if str.isalpha(message[i]):
        # Determine case and hanle accordingly
        if str.isupper(message[i]):
            newchar = chr((ord(message[i]) - ord('A') + shift) % 26 + ord('A'))
            print("{}".format(newchar), end = "")
            counter += 1
        else:
            newchar = chr((ord(message[i]) - ord('a') + shift) % 26 + ord('a'))
            print("{}".format(newchar), end = "")
            counter += 1
    # If character isn't alphabetical, print as is
    else:
        print("{}".format(message[i]), end = "")
    
# Return new line
print()
sys.exit(0)